### ISSUE TITLE
peerProfiles too large

### ISSUE THREAD

**Issue by khumarahn** [PurpleI2P/i2pd#1924]
Hi. I ran out space today: the peerProfiles folder grew over 2.5GB. This isn't that much, but I run i2pd on a cheap vps with very limited storage. I wonder if this is normal, and if there's a way to optimize storage.

**Comment by Vort** [comment]
I suggest you to shut down node, delete peerProfiles folder and start node again.

**Comment by khumarahn** [comment]
Oh, an attack?

**Comment by Vort** [comment]
Yes, network is under attack. There were several waves of spreading fake floodfills routerinfos.

**Comment by khumarahn** [comment]
Is there an automatic way to remove peer profiles, or it is only manual?

**Comment by Vort** [comment]
They are cleared automatically. But in case of attack there are just too many of them.

**Comment by SilmorSenedlen** [comment]
I have noticed several attacks on ~ past week, in addition to huge number of peer profile files, I almost got exhausted from memory.
Most of relays would probably have already fallen xD
![I2Pd_DoS_02_Res](https://user-images.githubusercontent.com/6761903/236584751-a6ad8cdd-a485-4b8c-9eaf-4da0c4836dcb.png)
![I2Pd_DoS_02](https://user-images.githubusercontent.com/6761903/236584749-96166fc8-491e-4995-8ef6-8fb3492dade1.png)

After I had to move the I2Pd working directory from the RAM disk to the HDD, I have the following picture:

![I2Pd_DoS_03](https://user-images.githubusercontent.com/6761903/236584879-dc46bfc3-9d9f-49ce-949c-5f782855f593.png)
And relative "small" memory usage: about 500 MiB.

**Comment by Vort** [comment]
@SilmorSenedlen that RAM consumption is a big problem. But I don't know how to track down why it happens.
From my experience, it can happen because of lags. Something somewhere make system overload and i2pd start glitching because of it.

First of all, you need to update your node.
I personally tested 2.47.0-67-gc1168c2a and it works relatively good.
But there are more commits made after it, will test them too soon.
If you will update node, then in the process it is better to clean `peerProfiles`.
As for `netDb`, I think it will clean itself in few hours after update.

**Comment by LLE8** [comment]
i2pd version 2.47.0-68-g2f74e670 (0.9.58)
floodfill = true
uptime 1 d 4 h
/var/lib/i2pd/peerProfiles/  folder grew from 0 to 2 GB
i2pd process memory usage: about 380 MB

**Comment by Vort** [comment]
For Windows, here:
https://github.com/PurpleI2P/i2pd/actions/workflows/build-windows.yml
https://github.com/PurpleI2P/i2pd/actions/runs/4903844731
https://github.com/PurpleI2P/i2pd/suites/12722999303/artifacts/683318292

**Comment by SilmorSenedlen** [comment]
> that RAM consumption is a big problem. But I don't know how to track down why it happens.
> From my experience, it can happen because of lags. Something somewhere make system overload and i2pd start glitching because of it.

Is it way to implement hard memory limit for i2pd, to save system and other apps from OOM(out of memory)?

**Comment by LLE8** [comment]
I think may be helpful
prlimit --as
prlimit --rss

**Comment by SilmorSenedlen** [comment]
> I think may be helpful
Sorry that I didn't indicate it right away, but I meant Windows*

**Comment by r4sas** [comment]
> > For Windows, here: https://github.com/PurpleI2P/i2pd/actions/workflows/build-windows.yml https://github.com/PurpleI2P/i2pd/actions/runs/4903844731 https://github.com/PurpleI2P/i2pd/suites/12722999303/artifacts/683318292
> 
> Thank you, but i need for linux (arm), some like this: i2pd_*bionic1_armhf.deb (maybe already have package?)

Unfortunately there is still no armhf/arm64 packages built from trunk. If anybody have packaging skills and know how OBS works: I need help with creating automated builds. [Here](https://build.opensuse.org/project/show/home:r4sas:i2pd) is working repository with release upload builds.

**Comment by r4sas** [comment]
> **r4sas**, не совсем понял в чем проблема - в создании пакета *.deb, или в компилировании бинарника?

В написании скриптов для OBS, которые бы позволили автоматически собирать пакеты для кода из транка.

**Comment by LLE8** [comment]
@snaike00 Raspberry Pi?

**Comment by LLE8** [comment]
Судя по "some like this: i2pd_*bionic1_armhf.deb" на этом устройстве Ubuntu 18.04?

**Comment by LLE8** [comment]
i2pd version 2.47.0-80-g623c3f46
Uptime 7 d 18 h
Size of /var/lib/i2pd/peerProfiles/ folder is 196 MB
i2pd process memory usage about 170 MB, 12000 transit tunnels

Либо атака прекратилась (временно), либо баг испрапвлен.

**Comment by SilmorSenedlen** [comment]
> Либо атака прекратилась (временно), либо баг испрапвлен.

I don't noticed any new attacks from past week.
And number of known Routers and Floodfills from my [previous post](https://github.com/PurpleI2P/i2pd/issues/1924#issuecomment-1536906116), it decreased to the current amount by itself, within a few days.
Version: **2.47.0**
```
Uptime: 11 days, 10 hours, 25 minutes, 32 seconds
Network status: OK
Network status v6: OK
Tunnel creation success rate: 10%
Received: 2103.02 GiB (947.92 KiB/s)
Sent: 2207.30 GiB (997.02 KiB/s)
Transit: 1947.43 GiB (895.81 KiB/s)
Routers: 10870 Floodfills: 1333 LeaseSets: 214
Client Tunnels: 27 Transit Tunnels: 8777
```
Virt. mem: 545 MiB
RAM: 400 MiB

**Comment by LLE8** [comment]
Значит там есть все нужные средства разработки и библиотеки? Может быть попробовать собрать непосредственно на самом устройстве? 512 МБ ОЗУ + 1ГБ свопа найдется? 
Отсюда https://i2pd.readthedocs.io/en/latest/devs/building/unix/  можно извлечь простую и работоспособную (проверено) инструкцию из десятка команд.
Русский вариант https://i2pd.readthedocs.io/ru/latest/build_notes_unix.html  кажется немного неправильный, вероятно устарел.

**Comment by r4sas** [comment]
Сейчас попробую включить сборку armhf/arm64 в PPA репозитории: https://launchpad.net/~purplei2p/+archive/ubuntu/i2pd-git/+packages

**Comment by LLE8** [comment]
А может для этого устройства есть OpenWrt?

**Comment by wekoq** [comment]
Maybe better to stop offtopic and close issue?

**Comment by LLE8** [comment]
Та же  2.47.0-80-g623c3f46 (0.9.58)
Uptime 201 d 21 h
/var/lib/i2pd/peerProfiles - 146 МБ и 36191 файл
В ОЗУ - 261200 КБайт
Думаю, можно закрывать.
